dbms/src/Flash/Coprocessor/DAGExpressionAnalyzer.cpp

```diff
@@ -1422,22 +1592,39 @@ String DAGExpressionAnalyzer::alignReturnType(
     return updated_name;
 }
 
+template <bool need_col_ptr>
 void DAGExpressionAnalyzer::initChain(ExpressionActionsChain & chain) const
 {
     if (chain.steps.empty())
     {
         const auto & columns = getCurrentInputColumns();
-        NamesAndTypesList column_list;
         std::unordered_set<String> column_name_set;
-        for (const auto & col : columns)
+        if constexpr (need_col_ptr)
```

Comment:
Some hack...
In TiFlash, the sample block does not have a column pointer, while the header does have the block.
And here, we are setting a sample block for ExpressionAction.

Reply:
> 

However, in order to get the data type at compilation stage we need this column pointer.

Reply:
DataTypePtr is enough, all type-related judgments can be done with DataTypePtr

Reply:
`need_col_ptr` is still needed now?

Reply:
> `need_col_ptr` is still needed now?

okk, I have deleted this now.